Mathlib/Tactic/ApplyFun.lean

```diff
@@ -64,28 +64,47 @@ def applyFunHyp (f : TSyntax `term) (using? : Option Expr) (h : FVarId) (g : MVa
 def applyFunTargetFailure (f : Expr) : MetaM (List MVarId) := do
   throwError "`apply_fun` could not apply `{f}` to the main goal."
 
+/-- Coerce `f` to be a function, if needed. -/
+def ensureFun (f : Expr) : TacticM Expr := do
+  let ma ← mkFreshTypeMVar
+  let mb ← mkFreshTypeMVar
+  let ty := Expr.forallE `a ma mb .default
+  runTermElab <| Term.ensureHasType ty f
+
 /-- Apply a function to the main goal. -/
-def applyFunTarget (f : Expr) (using? : Option Expr) (g : MVarId) : MetaM (List MVarId) := do
+def applyFunTarget (f : Expr) (using? : Option Expr) (g : MVarId) : TacticM (List MVarId) := do
   match (← g.getType).getAppFnArgs with
-  | (``Ne, #[_, _, _]) => g.apply (← mkAppM ``ne_of_apply_ne #[f])
+  | (``Ne, #[_, _, _]) => g.apply (← mkAppM ``ne_of_apply_ne #[← ensureFun f])
```

Comment:
Can we use `Qq` here and let the elaborator deal with inserting coercions? I think we essentially did that in Lean 3.

Cc @gebner

Reply:
Doesn't qq elaborate at read time, not at run time? It shouldn't be able to make use of coercions if f ends up with the wrong type.

I could elaborate an expression with f substituted in using syntax quotations, but this change seemed epsilon easier. Ideally it would make use of the expected type too, but this is an improvement as I was just passing by.

Reply:
Hmm, `ensureFun` ended up being wrong (it created metavariables that never got solved for). Now it's using `Term.elabTermEnsuringType` and taking into account the expected type, as well as trying to solve for those metavariables.

There's a small issue that if you use holes (like `apply_fun f ?_`) it doesn't create new goals for the holes. This was already an issue, but the function that is most likely to be able to help (`withCollectingNewGoalsFrom`) can only return a single expression.

Reply:
Could we work around this, by instead of constructing the pair of expressions `(g', ginj)`, just build a pair containing both, all inside `withCollectingNewGoalsFrom`, and then immediate decompose the pair expression to finish the function?

Reply:
I'll mark as `awaiting-author` for now, but I don't mind whether you either try this hack, or just add a comment somewhere to the effect that `apply_fun f ?_` will forget the additional goals.

Reply:
@semorrison That's a fun idea for a hack. I found another solution, where you create some metavariables outside `withCollectingNewGoalsFrom` and then assign them from within. There were two things that ended up needing some care: (1) you need to return an expression that definitely contains all the new metavariables from placeholders (and it can't just be a metavariable that has a type that contains these metavariables!) and (2) you need to avoid assigning the goal metavariable, since `withCollectingNewGoalsFrom` uses `getMainTag` to name anonymous placeholders.

Reply:
Nice!